### PR TITLE
Allow disabling scrolling in Tree and implement horizontal scrolling

### LINKED ISSUE #1793: Add a horizontal scrollbar to the Scene tree dock
**Describe the project you are working on:**
I'm currently working on a city builder game.

**Describe the problem or limitation you are having in your project:**
The other designer made a UI mock up but I am unable to select all the nodes. As shown in the image below, I simply cannot see most of the children. 
image

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Adding a horizontal slider at the bottom would allow me to move the view in the window to the right so I can select the desired nodes.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
image

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
The way to work around this issue is making a lot of different scenes. This would require us to save all the components of the UI before we actually know if we want to keep them. 

**Is there a reason why this should be core and not an add-on in the asset library?:**
I'm not sure. it could be an add-on? Just feels a bit weird if we'd have to download one for this type of issue.

### LINKED ISSUE #2357: Add HSlider to Tree
### Describe the project you are working on
Plugins

### Describe the problem or limitation you are having in your project
I have a Tree node with dynamic content and I want to scroll it horizontally.

I have tried to put it in `ScrollContainer` but I have 2 problems:

1. I need to calculate "actual" width of the tree manually (ideally taking into account all parameters, including the theme ones) and also automatically update it as needed. It doesn't seem like a trivial task.

2. I have some problems with scrolling, for example there seems to be a conflict between horizontal and vertical sliders (when using a ScrolContainer).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Adding `HSlider` to `Tree` would solve the problem

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

https://user-images.githubusercontent.com/16458555/109239398-2c5f4280-77f7-11eb-9558-08667427ea1b.mp4

- This horizontal slider might be optional.
In such a case, a `horizontal_enabled` property could be added.
- Maybe some other properties that ScrollContainer has could be added (like `h_value`)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Could be work-rounded, but not with a few lines of script.

### Is there a reason why this should be core and not an add-on in the asset library?
this could be used often when creating responsive UIs without the hassle and reinventing the wheel.
Imo this would be a logical addition to the existing vertical scroll slider.

### LINKED ISSUE #2571: Allow disabling scrolling in Tree
### Describe the project you are working on

Dialog tree plugin for the Godot Engine.

### Describe the problem or limitation you are having in your project

I use `Tree` nodes to display a list of options, but if I don't get the size exactly right, a scrollbar appears and makes the tree scrollable.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Either a simple `enable_scrolling` boolean or `get_h / v_scrollbar()` functions like that of `ScrollContaine` added to Tree.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```gdscript
enable_scrolling : bool
get_h_scrollbar() : HScrollBar
get_v_scrollbar() : VScrollBar
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You can hide the scrollbar by making it invisible with a theme and reset the scroll value every frame, which is very hacky.

### Is there a reason why this should be core and not an add-on in the asset library?

It's a core node.

### PR DESCRIPTION
**Pull request by groud**
Similar to what's done in ScrollContainer, this adds a way to disable scroling for the Tree node. In this case, it has a minimum size. I need it for the TileSet editor rework. See how it works:
Peek 25-06-2021 21-34

Edit: Also, this PR now also fix the computation of the internal minimum width. It also renames `set_column_min_width` to `set_column_custom_minimum_width`.

Closes godotengine/godot-proposals#2571.
Closes https://github.com/godotengine/godot-proposals/issues/2357
Closes https://github.com/godotengine/godot-proposals/issues/1793

### CODE REVIEW

**Comment by pycbouh** [comment]
This seem to close https://github.com/godotengine/godot-proposals/issues/2571.

And do I understand correctly that this also implements https://github.com/godotengine/godot-proposals/issues/2357? And if so, can you look into fixing https://github.com/godotengine/godot-proposals/issues/1793 along the way?

File: scene/gui/tree.cpp [hunk 1]
```diff
@@ -3513,7 +3525,17 @@ void Tree::_update_all() {
 }
 
 Size2 Tree::get_minimum_size() const {
-	return Size2(1, 1);
+	if (h_scroll_enabled && v_scroll_enabled) {
+		return Size2();
```
Comment (Macksaur) [inline scene/gui/tree.cpp:3529]: What happened to `Size2(1, 1)`? Does this change any behaviour elsewhere unexpectedly? Or is this desired?

Comment (groud) [reply scene/gui/tree.cpp:3529]: This is desired, it does not make any sense to keep a minimum size of (1, 1)

Comment (Macksaur) [reply scene/gui/tree.cpp:3529]: I understand, I am just worried about divisions by zero elsewhere in the code that has relied on this previous minimum size behaviour or some such.

Comment (groud) [reply scene/gui/tree.cpp:3529]: That's unlikely, a lot of other Control nodes return Size2() when they do not have a minimum size. Which is the case here.

**Comment by groud** [comment]
> And do I understand correctly that this also implements godotengine/godot-proposals#2357? And if so, can you look into fixing godotengine/godot-proposals#1793 along the way?

Alright, I pushed another commit to the PR so that this works.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**